### PR TITLE
Includes container with fetch on SSR

### PR DESCRIPTION
**Pull request by rhenshaw56**
Resolves https://github.com/bbc/simorgh-infrastructure/issues/898

**Overall change:** 
_Adds container for includes._

**Code changes:**
- Created includes container
- Added toggle for Includes
- Fetch include markup 
- add constraints that limit rendering markup
- added timeout option to fetch request

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added labels to this PR for the relevant pod(s) affected by these changes
- [X] I have assigned this PR to the Simorgh project

**Testing:**
- [X] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [X] This PR requires manual testing

### CODE REVIEW

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: I'm a bit unsure where the encoding is necessary? I thought something in the express middleware would be clever enough to do this for you?

Looks like you may have missed single quotes (`'`) btw: https://stackoverflow.com/questions/40263803/native-javascript-or-es6-way-to-encode-and-decode-html-entities

**Review by andrewscfc — COMMENTED** [review]
Just a quick review on the draft, I'm sure you are already aware that the tests need updating as the fetch has moved.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: > Looks like you may have missed single quotes (') btw: https://stackoverflow.com/questions/40263803/native-javascript-or-es6-way-to-encode-and-decode-html-entities

Thanks for this.

> I'm a bit unsure where the encoding is necessary?

Passing the html text from the server distorts the hydration as the text contains un-escaped literals. It would render but the markup will also appear at the bottom of the page and any property attached on the `window` object on the server becomes `undefined`.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (jamesdonoh) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:24]: Can we revisit the need for this conversion and check if there is any alternative?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (jamesdonoh) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:46]: Do we need to indent?

Comment (jamesdonoh) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:46]: Can we move this to the logger?

File: src/app/containers/Include/index.jsx
Comment (simonsinclair) [inline src/app/containers/Include/index.jsx:43]: Investigate if this is indeed necessary. I think we should understand why this component is rendering differently on the server to the client.

File: src/app/containers/Include/index.test.jsx
Comment (simonsinclair) [inline src/app/containers/Include/index.test.jsx]: Should this be `true`?

File: src/app/containers/Include/index.test.jsx
Comment (simonsinclair) [inline src/app/containers/Include/index.test.jsx]: Should this be `true`?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (simonsinclair) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:36]: Would it make sense to pass the response text straight to the component?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (simonsinclair) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: When/if this is made client-side, will 2 seconds be long enough?

Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: at some point 2 was timing out, changed to 3seconds

File: src/app/containers/Include/index.jsx
Comment (rhenshaw56) [reply src/app/containers/Include/index.jsx:43]: this is needed as we're still decoding the markup

File: src/app/containers/Include/index.jsx
Comment (andrewscfc) [inline src/app/containers/Include/index.jsx:11]: Could we add a comment here to quickly explain why the decode is necessary? Just explaining that the data is inlined into the page here https://github.com/bbc/simorgh/blob/62184ec4a7638eaee261b85c68e70f76e4510bc2/src/server/Document/component.jsx#L82 and needs to be escaped to prevent the page being corrupted.

Just makes it clear why we have this logic if we stick with this approach longer term.

File: src/app/containers/Include/index.jsx
Comment (andrewscfc) [reply src/app/containers/Include/index.jsx:43]: Maybe another comment to explain this is needed for the same reason above

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:24]: Similar comment to explain the need for the encode as mentioned above

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:36]: I tried this and it didn't have the desired effect due to encoding issues we've all dug into

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: Happy to see this timeout, I think it's important to have it so we don't exhaust the sockets on our simorgh EC2 boxes if the smallprox service fails to respond quickly.

Did you ever confirm if the timeout applies on the client @rhenshaw56 with `isomorphic-fetch`? We wouldn't want this timeout in the browser, we need to allow includes to take to time to resolve due to local network conditions.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:46]: Did we come to a consensus on this? I believe a similar logger approach was mentioned for toggles where an indent was used.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]:
```suggestion
      // `includeUrls[includeType]` here should be replaced with `href` once mozart routes have been created. /*TODO: Create issue for this */
```
I thought an issue was created for this?

File: src/app/containers/Include/index.jsx
Comment (simonsinclair) [reply src/app/containers/Include/index.jsx:43]: What happens if hydration is delayed or doesn't occur? Is the Include or page broken? Hydration could be delayed or not occur for multiple reasons. E.G. If the user is on a slow connection and client-side JS hasn't executed yet, or JS is disabled.

Edit: Answers have been added in [this comment](https://github.com/bbc/simorgh/pull/5894#issuecomment-603842890).

**Review by simonsinclair — COMMENTED** [review]
Reviewed again, as requested. Please see updated comment.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: I don't think these consts need to be exported?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: We are using the same code to generate the expected markup as is used in `convertInclude` itself. Should we hardcode the encoded markup in a _set of new consts_ at the top of this file? That way if we change the encoding logic later we would catch any regressions e.g. we accidently didn't encode

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:190]: Should assert fetch hasn't been called here: `expect(fetch).not.toHaveBeenCalled();`

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]:
```suggestion
// this defaults all VJ, IDT2 and IDT1 includes to the specified urls for now
```

**Review by andrewscfc — COMMENTED** [review]
All fairly minor stuff, thanks for sticking with this 👍 

I'd rather not approve this PR as I've contributed to it to some degree, best to get approvals from other team members.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: yeah an issue was created, will edit the comment

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: yeah, makes sense

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:190]: it shouldn't. will remove it

File: src/app/containers/Include/index.jsx
Comment (12) [reply src/app/containers/Include/index.jsx:11]: As someone who isn't all too familiar with this approach to includes, a description would definitely help here.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (12) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: Has this timeout been confirmed with UX? They might want some input into how long we/the user should be waiting.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:46]: @andrewscfc I think we agreed it was okay to have in making logging more illustrative

**Comment by rhenshaw56** [comment]
>What happens if hydration is delayed or doesn't occur? Is the Include or page broken? Hydration could be delayed or not occur for multiple reasons. E.G. If the user is on a slow connection and client-side JS hasn't executed yet, or JS is disabled.

@simonsinclair tested this with JS disabled and the Includes render and displays for IDT2 content but for VJ's it shows `Please Upgrade Your browser` (which we expect because the VJ markup has scripts within it). Tried this out again for a PAL page with JS disabled for the same piece of VJ content which we currently have and it has the same behaviour.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (simonsinclair) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: To add to Rhys' comment - what happens if the request times out? Do we notify the user?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (HarveyPeachey) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:7]: Can we add the PR number which resolves this TODO? https://github.com/bbc/simorgh-infrastructure/issues/860

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: @12 @simonsinclair at the moment we just return a null value so it never renders. Good point on clarifying with UX.

File: src/app/containers/Include/index.jsx
Comment (HarveyPeachey) [inline src/app/containers/Include/index.jsx:42]: @greenc05 Do we need to have anything a11y specific for include containers?

**Review by HarveyPeachey — COMMENTED** [review]
Great work!

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (12) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: I think perhaps something to indicate that the content is unavailable (similar to social embeds) would be a better solution, but that's UX territory.

File: src/app/containers/Include/index.test.jsx
Comment (12) [inline src/app/containers/Include/index.test.jsx]:
```suggestion
const fakeMarkup = `&lt;div&gt;Visual Journalism Markup&lt;/div&gt;&lt;script type=&quot;text/javascript&quot; src=&quot;localhost/vj.js&quot;&gt;&lt;/script&gt;`;
```

I know it's only fixture data 🤷‍♂

File: src/app/containers/Include/index.jsx
Comment (rhenshaw56) [reply src/app/containers/Include/index.jsx:42]: not at the moment @HarveyPeachey

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: I'm pretty sure we confirmed we had no fallback UX to implement when we started work on this @lizcameron @jimjohnsonrollings , can you confirm this?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:46]: okay @jamesdonoh , can we resolve this?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:190]: > it shouldn't. will remove it

I was suggesting adding the extra `expect(fetch).not.toHaveBeenCalled();` because for a non-supported include type no fetch will be made.

File: src/app/containers/Include/index.jsx
Comment (andrewscfc) [reply src/app/containers/Include/index.jsx:11]: Thanks for adding the comment on encode @rhenshaw56 , are still going add one here to explain the need to decode?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (lizcameron) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:31]: Yep, that's right @andrewscfc - we don't have UX for a fallback at the moment. If the include doesn't load/times out, then it just doesn't show on the page.

File: src/app/containers/Include/index.jsx
Comment (andrewscfc) [reply src/app/containers/Include/index.jsx:11]: Cheers!

**Review by andrewscfc — COMMENTED** [review]
Thanks for all the hard work @rhenshaw56 , all my comments are addressed 💯 

I won't approve myself but happy to take this to test from my perspective 🙂

**Review by HarveyPeachey — APPROVED** [review]
LGTM! Following the PR number addition

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (chris-hinds) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:39]: Can we log the URL we tried to fetch in this event. It would make debugging a lot easier.

Comment (chris-hinds) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:39]: Also the event should probably be namespaced. Something like `includes_fetch_error` or `smallprox_fetch_error` `http_fetch_error` is very generic.

**Review by chris-hinds — COMMENTED** [review]
Nice work so far, just two small comments from me.

**Review by chris-hinds — APPROVED** [review]

**Review by simonsinclair — APPROVED** [review]
This is a great piece of work. Thanks for adding a comment explaining the fetch timeout. 👍

**Comment by paruchurisilpa** [comment]
Looks good to me..

**Pull request merged**
